onnxruntime/core/providers/coreml/builders/impl/pad_op_builder.cc

```diff
@@ -0,0 +1,215 @@
+    LOGS(logger, VERBOSE) << "Pad requires input shape to be at least 2, input is "
+                          << input_shape.size() << "d shape";
+    return false;
+  }
+
+  if (std::find(input_shape.begin(), input_shape.end(), int64_t{0}) != input_shape.end()) {
+    LOGS(logger, VERBOSE) << "Input with a 0 for a dimension is not supported";
+    return false;
+  }
+
+  {
+    NodeAttrHelper helper(node);
+    const auto mode = helper.Get("mode", "constant");
+    if (mode != "constant") {
+      LOGS(logger, VERBOSE) << "Only `constant` mode Pad is currently supported for now, mode: " << mode;
+      return false;
+    }
+
+    if (input_defs.size() < 3) {
+      LOGS(logger, VERBOSE) << "`constant_value` input is required for constant mode Pad op.";
+      return false;
+    }
+  }
+
+  // only support if `pads` input is known and does not contain negative values and only applies padding values
+  // for last two dimensions.
+  {
+    const auto pads_initializer_it = initializers.find(input_defs[1]->Name());
+    if (pads_initializer_it == initializers.end()) {
+      LOGS(logger, VERBOSE) << "pads must be a constant intializer.";
+      return false;
+    }
+
+    const ONNX_NAMESPACE::TensorProto& pads_initializer = *pads_initializer_it->second;
+    Initializer unpacked_tensor(pads_initializer);
+
+    auto pads_tensor_data = unpacked_tensor.DataAsSpan<int64_t>();
+    for (int64_t i = 0; i < unpacked_tensor.size(); i++) {
+      if (pads_tensor_data[i] < 0) {
+        LOGS(logger, VERBOSE) << "Negative pad value is not supported: pads["
+                              << i << "] = " << pads_tensor_data[i];
+        return false;
+      }
+    }
+
+    // Check that only supports padding on last two dimensions - [H,W].
+    // CoreML PaddinglayerParams: https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#paddinglayerparams
+    const auto input_rank = input_shape.size();
+    // if not provided, make a default axes input as [0, 1, ..., input_rank - 1]
+    std::vector<int64_t> default_axes(input_rank);
+    std::iota(std::begin(default_axes), std::end(default_axes), 0);
+    auto axes_tensor_data = gsl::span<const int64_t>(default_axes);
+    if (input_defs.size() > 3) {
+      // optional axes input is provided
+      const auto axes_initializer_it = initializers.find(input_defs[3]->Name());
+      if (axes_initializer_it == initializers.end()) {
+        LOGS(logger, VERBOSE) << "If provided, axes must be a constant intializer.";
+        return false;
+      }
+      const ONNX_NAMESPACE::TensorProto& axes_initializer = *axes_initializer_it->second;
+      Initializer unpacked_tensor(axes_initializer);
+      axes_tensor_data = unpacked_tensor.DataAsSpan<int64_t>();
+    }
+    int64_t num_axes = axes_tensor_data.size();
+    for (int64_t i = 0; i < num_axes; i++) {
+      if (SafeInt<int64_t>(axes_tensor_data[i]) < input_rank - 2) {
+        if (pads_tensor_data[i] != 0 || pads_tensor_data[i + num_axes] != 0) {
```

Comment:
should it be using `axes_tensor_data[i]`?
```suggestion
        if (pads_tensor_data[axis_tensor_data[i]] != 0 || pads_tensor_data[axis_tensor_data[i] + num_axes] != 0) {
```